### PR TITLE
Remove copy of tifffile; install from pip. Fixes #3578

### LINKED ISSUE #3578: Check if tifffile can be added as a requirement instead of vendoring it
## Description

So far, we have a modified copy of tifffile. See in particular #3452 

If wheels are available one day for tifffile, then we must consider removing this copy.

### PR DESCRIPTION
**Pull request by kalvdans**
Fixes https://github.com/scikit-image/scikit-image/issues/3578

- [x] Add failing test for LZW compressed TIFF files.
- [x] Add `imagecodecs-lite` as pip requirement to fix above test failure.

## For reviewers

- [ ] Check that the PR title is short, concise, and will make sense 1 year
  later.
- [ ] Check that new functions are imported in corresponding `__init__.py`.
- [ ] Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.
- [ ] Consider backporting the PR with `@meeseeksdev backport to v0.14.x`

### CODE REVIEW

**Comment by kalvdans** [comment]
I would need some help understanding why skimage/future/graph/tests/test_rag.py fails in my branch, it looks unrelated to my changes.

**Comment by hmaarrfk** [comment]
is it only in your branch?

**Comment by jarrodmillman** [comment]
I released a new version of networkx (2.4rc1) yesterday.  We removed some code that was deprecated in 2.0.  I am looking into it now.  (I only looked at the error in continuous-integration.scikit-image (Default Python37-x64-pre).)

File: skimage/io/_plugins/tifffile_plugin.py [hunk 1]
```diff
@@ -1,9 +1,6 @@
 from warnings import warn
 
-try:
-    from tifffile import TiffFile, imsave, parse_kwargs
-except ImportError:
-    from ...external.tifffile import TiffFile, imsave, parse_kwargs
+from tifffile import TiffFile, imsave, parse_kwargs
 
 
 def imread(fname, dtype=None, **kwargs):
```
Comment (hmaarrfk) [inline skimage/io/_plugins/tifffile_plugin.py]: There might be some behaviour here that we want to deprecate. We can leave that for an other PR though.

**Comment by jni** [comment]
We should also depend on [imagecodecs-lite](https://pypi.org/project/imagecodecs-lite/), as most tiffs have LZW compression that the pure-python tifffile does not support.

**Comment by kalvdans** [comment]
> We should also depend on [imagecodecs-lite](https://pypi.org/project/imagecodecs-lite/), as most tiffs have LZW compression that the pure-python tifffile does not support.

Yes, I saw it mentioned in the issue. Can you write a test for that in `skimage/io/tests/test_tifffile.py`?

**Comment by kalvdans** [comment]
> is it only in your branch?

It seems scikit-image doesn't freeze its dependencies so external changes will hit all innocent branches. I personally prefer frozen dependencies but that's another story. Thanks for dealing with it so quickly in https://github.com/scikit-image/scikit-image/pull/4236 ! I will rebase and forcepush this pull request now.

**Comment by jni** [comment]
@kalvdans could you make it `imagecodecs-lite` rather than `imagecodecs`? The former is a much easier dependency to install.

Re: testing, if I understand correctly, it should be sufficient to do a round-trip write/read with `compress=1`.

**Comment by kalvdans** [comment]
@jni
Seems I can't get any more tests to function with imagecodecs-lite. See matrix below.
```
,--------------------+--------------+-------------+--------------+-------------,
| packages installed | zlib imwrite | zlib imread | lzma imwrite | lzma imread |
+--------------------+--------------+-------------+--------------+-------------+
| none               |      Y       |     Y       |     N        |     N       |
| imagecodecs-lite   |      Y       |     Y       |     N        |     N       |
| imagecodecs        |      Y       |     Y       |     Y        |     Y       |
'--------------------+--------------+-------------+--------------+-------------'
```

**Comment by jni** [comment]
The codecs supported by `imagecodecs-lite` [are](https://pypi.org/project/imagecodecs-lite/):

- LZW, PackBits, Delta, XOR Delta, Packed Integers, Floating Point Predictor, and Bitorder reversal

LZW in particular is one that I've seen used in various files in the wild.

**Comment by kalvdans** [comment]
Any idea how to create a LZW-compressed image with tifffile?

**Comment by jni** [comment]
@kalvdans I thought I knew but apparently not. @cgohlke any idea what is happening here? I have `imagecodecs-lite` installed...

```python
In [2]: import tifffile                                                                                                                                
In [4]: import numpy as np                                                                                                                             
In [6]: image = np.random.randint(0, 2**16-1, size=(2048, 2048))                                                                                       
In [8]: tifffile.imsave('test.tif', image, compress=('lzw', 1))                                                                                        
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-8-2a949f858cb9> in <module>
----> 1 tifffile.imsave('test.tif', image, compress=('lzw', 1))

~/miniconda3/envs/cf/lib/python3.6/site-packages/tifffile/tifffile.py in imwrite(file, data, shape, dtype, **kwargs)
    785 
    786     with TiffWriter(file, **tifargs) as tif:
--> 787         return tif.save(data, shape, dtype, **kwargs)
    788 
    789 

~/miniconda3/envs/cf/lib/python3.6/site-packages/tifffile/tifffile.py in save(self, data, shape, dtype, returnoffset, photometric, planarconfig, extrasamples, tile, contiguous, align, truncate, compress, rowsperstrip, predictor, subsampling, colormap, description, datetime, resolution, subfiletype, software, metadata, ijmetadata, extratags)
   1766         # define compress function
   1767         if compress:
-> 1768             compressor = TIFF.COMPESSORS[compresstag]
   1769             if predictor:
   1770 

KeyError: <COMPRESSION.LZW: 5>

In [9]: tifffile.imsave('test.tif', image, compress='lzw')                                                                                             
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-9-50aa306ce580> in <module>
----> 1 tifffile.imsave('test.tif', image, compress='lzw')

~/miniconda3/envs/cf/lib/python3.6/site-packages/tifffile/tifffile.py in imwrite(file, data, shape, dtype, **kwargs)
    785 
    786     with TiffWriter(file, **tifargs) as tif:
--> 787         return tif.save(data, shape, dtype, **kwargs)
    788 
    789 

~/miniconda3/envs/cf/lib/python3.6/site-packages/tifffile/tifffile.py in save(self, data, shape, dtype, returnoffset, photometric, planarconfig, extrasamples, tile, contiguous, align, truncate, compress, rowsperstrip, predictor, subsampling, colormap, description, datetime, resolution, subfiletype, software, metadata, ijmetadata, extratags)
   1766         # define compress function
   1767         if compress:
-> 1768             compressor = TIFF.COMPESSORS[compresstag]
   1769             if predictor:
   1770 

KeyError: <COMPRESSION.LZW: 5>
```

```
 $ pip list | grep tifffile
tifffile                      2019.7.2                
 $ pip list | grep imagecodecs-lite
imagecodecs-lite              2019.4.20               
```

**Comment by jni** [comment]
(Note: adding `.astype(np.uint16)` to the image creation line does nothing useful.)

**Comment by kalvdans** [comment]
I created a small lzw test file in gimp and put it in the data folder. I think it is good to merge now, unless we want to add some change log.

**Comment by kalvdans** [comment]
Anyone volunteering to repair the Travis CI?

**Comment by hmaarrfk** [comment]
other minimum dependencies seem to need upgrading. 

Seems like image codecs lite doesn't have wheels for linux or mac osx

https://pypi.org/project/imagecodecs-lite/2019.4.20/#files

Depending on this would not be good.

**Comment by kalvdans** [comment]
Ok, I'm back to the state before adding new tests. Feel free to merge or close, but if you merge, squash the commits!

**Comment by hmaarrfk** [comment]
lets see what @jni has to say.

**Comment by jni** [comment]
Hey guys, sorry, a bit swamped at the moment, will need to think about what's best here, particularly testing before/after with a few tiffs from local labs. Thank you @kalvdans for your work here, sorry for the back-and-forth!

**Comment by sciunto** [comment]
This PR is definitely a good thing. I'm +1 to install tifffile from pip.

**Comment by alexdesiqueira** [comment]
+1 from me as well; I'll wait for @jni too. Thank you for your work so far, @kalvdans!

**Comment by stefanv** [comment]
It does't look as though @jni's concern has been addressed, nor has @cgohlke looked into it, so let's hold off until we can figure out how to support those compressed TIFFs.

**Comment by cgohlke** [comment]
> It does't look as though @jni's concern has been addressed, nor has @cgohlke looked into it,

I assume this is regarding https://github.com/scikit-image/scikit-image/pull/4235#issuecomment-541538692

The reason is that imagecodecs does not include a LZW encoder, only a decoder. I did not have the time and incentive to implement it yet.

Btw, `imagecodecs-lite` is no longer used by tifffile. `imagecodecs` now provides nearly full-featured multilinux and macOS wheels and allows to skip unwanted or troublesome codecs when building from source.

**Comment by stefanv** [comment]
> Btw, `imagecodecs-lite` is no longer used by tifffile. `imagecodecs` now provides nearly full-featured multilinux and macOS wheels and allows to skip unwanted or troublesome codecs when building from source.

That's very helpful to know.  So looks like we can just make that a default dependency, then we're good to go.

**Comment by hmaarrfk** [comment]
is this good to merge today?

**Comment by jni** [comment]
@hmaarrfk that last comment from @cgohlke is certainly awesome! =D Please let me try this all out and I'll merge ASAP.

**Comment by hmaarrfk** [comment]
just, i'm just keeping this alive ;)

**Comment by jni** [comment]
Absolutely awesome. Tested this locally and it works a treat. Thank you so much @kalvdans for this contribution and @cgohlke for your tireless work on tifffile! I'm over the moon to merge this!

**Comment by alexdesiqueira** [comment]
Thank you for your work @kalvdans! Really appreciated :)

**Comment by stefanv** [comment]
:rocket:

**Pull request merged**